File: android/app/src/main/kotlin/com/cruisemesh/app/friending/FriendRequestSender.kt
package com.cruisemesh.app.friending

import android.content.Context
import android.util.Log
import com.cruisemesh.app.chat.UserIdHex
import com.cruisemesh.app.identity.ProfileStore
import com.cruisemesh.app.mesh.MeshRouter
import com.cruisemesh.app.mesh.RelaySyncEvents
import com.cruisemesh.app.relay.RelayConfigStore
import uniffi.cruisemesh_core.Contact
import uniffi.cruisemesh_core.Identity
import uniffi.cruisemesh_core.MessageStore
import uniffi.cruisemesh_core.SharedFriendCard
import uniffi.cruisemesh_core.makeFriendCard
import uniffi.cruisemesh_core.makeSharedFriendRequestPayload

private const val TAG = "FriendRequestSender"

/**
 * Queues the mutual-friending follow-up from DESIGN.md §6.2: once we scan a
 * peer's QR card and import them locally, send our own signed friend card
 * back as a hidden `kind=3` chat-stream message so they can auto-import us.
 */
object FriendRequestSender {

    fun queueForScannedContact(
        context: Context,
        store: MessageStore,
        identity: Identity,
        contact: Contact,
    ): FriendRequestDelivery = queue(context, store, identity, contact, shared = null)

    /**
     * The same mutual `kind=3`, carrying the shared card it came from as the
     * optional tail (specs/share-contact.md). The tail is what makes the other
     * phone hold the request for confirmation instead of auto-importing, so it
     * must ride on exactly the request a shared-card import sends back.
     */
    fun queueForSharedCard(
        context: Context,
        store: MessageStore,
        identity: Identity,
        contact: Contact,
        shared: SharedFriendCard,
    ): FriendRequestDelivery = queue(context, store, identity, contact, shared)

    private fun queue(
        context: Context,
        store: MessageStore,
        identity: Identity,
        contact: Contact,
        shared: SharedFriendCard?,
    ): FriendRequestDelivery {
        val relay = RelayConfigStore.load(context)
        val payload = try {
            val cardJson = makeFriendCard(
                ProfileStore.loadDisplayName(context),
                identity,
                relay?.relayUrl,
                relay?.relayToken,
            )
            if (shared == null) cardJson else makeSharedFriendRequestPayload(cardJson, shared)
        } catch (e: Exception) {
            Log.w(TAG, "Skipping invalid friend card", e)
            return FriendRequestDelivery(reachedDirectly = false, lamport = 0uL)
        }
        val timestamp = System.currentTimeMillis()
        val authored = try {
            store.authorFriendRequest(identity, contact, payload, timestamp)
        } catch (e: Exception) {
            return FriendRequestDelivery(reachedDirectly = false, lamport = 0uL)
        }
        RelaySyncEvents.requestSync()

        val reachedDirectly = MeshRouter.sendToUserId(contact.userId, authored.frame)
        if (!reachedDirectly) {
            val muled = MeshRouter.relayToAll(authored.frame)
            Log.i(
                TAG,
                "Queued friend request for ${UserIdHex.encode(contact.userId)}; " +
                    "peer not currently connected, sprayed to $muled mule link(s)",
            )
        }
        return FriendRequestDelivery(reachedDirectly = reachedDirectly, lamport = authored.message.lamport)
    }
}

File: ios/CruiseMesh/Friending/FriendRequestSender.swift
import Foundation
import os.log

enum FriendRequestSender {
    private static let log = Logger(subsystem: "com.cruisemesh", category: "FriendRequest")

    /// After importing a scanned friend card, queue a signed kind=3 back.
    ///
    /// `shared` is set only when the card we just imported came out of somebody
    /// else's **Share contact** code (specs/share-contact.md). It rides along as
    /// a backwards-decodable tail so the other phone can tell an introduction
    /// from a scan and ask before importing; everything else about the send is
    /// identical, which is why this is one parameter rather than a second path.
    @discardableResult static func sendMutualFriendRequest(
        store: MessageStore,
        identity: Identity,
        contact: Contact,
        displayName: String,
        shared: SharedFriendCard? = nil
    ) -> FriendRequestDelivery {
        guard let ownCardJson = try? makeFriendCard(
            name: displayName.isEmpty ? "Friend" : displayName,
            identity: identity,
            relayUrl: RelayConfigStore.load()?.relayUrl,
            relayToken: RelayConfigStore.load()?.relayToken
        ) else {
            return FriendRequestDelivery(reachedDirectly: false, lamport: 0)
        }
        let cardJson: String
        if let shared {
            // Fail rather than silently sending a tailless request: that would
            // land as an auto-import on their phone, which is exactly the
            // confirmation this flow exists to ask for.
            guard let tailed = try? makeSharedFriendRequestPayload(
                cardJson: ownCardJson,
                shared: shared
            ) else {
                return FriendRequestDelivery(reachedDirectly: false, lamport: 0)
            }
            cardJson = tailed
        } else {
            cardJson = ownCardJson
        }
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        guard let authored = try? store.authorFriendRequest(
            identity: identity, contact: contact, friendCardJson: cardJson, timestampMs: timestamp
        ) else {
            return FriendRequestDelivery(reachedDirectly: false, lamport: 0)
        }
        ChatEvents.notifyChatChanged(authored.message.chatId)
        RelaySyncEvents.requestSync()
        let reachedDirectly = MeshRouter.sendToUserId(userId: contact.userId, frame: authored.frame)
        if !reachedDirectly {
            let muled = MeshRouter.relayToAll(frame: authored.frame)
            log.info("Friend request queued for later delivery to \(UserIdHex.encode(contact.userId), privacy: .public); sprayed to \(muled) mule link(s)")
        }
        return FriendRequestDelivery(reachedDirectly: reachedDirectly, lamport: authored.message.lamport)
    }
}

struct FriendRequestDelivery: Hashable {
    let reachedDirectly: Bool
    let lamport: UInt64
}
